File: coco-sdk/python/src/coco_sdk/runtime.py
"""Runtime discovery helpers for locating the ``coco`` CLI binary."""

from __future__ import annotations

import os
import shutil
from dataclasses import dataclass
from pathlib import Path
from typing import Mapping

from coco_sdk.errors import CLINotFoundError


@dataclass(frozen=True)
class CocoRuntime:
    """Resolved coco runtime executable.

    ``source`` is diagnostic-only and helps tests or callers explain
    whether the binary came from an explicit argument, ``COCO_PATH``,
    ``PATH``, or a well-known local install location.
    """

    binary_path: str
    source: str


def resolve_coco_runtime(
    binary_path: str | os.PathLike[str] | None = None,
    *,
    env: Mapping[str, str] | None = None,
) -> CocoRuntime:
    """Resolve the coco CLI runtime executable.

    This is intentionally lightweight: today it discovers an already
    installed binary. Keeping the logic behind one public seam lets a
    future platform-specific runtime package plug in without changing
    the subprocess transport or e2e harness.
    """

    environment = env if env is not None else os.environ
    if binary_path is not None:
        return _runtime_from_candidate(Path(binary_path), "explicit")

    env_path = environment.get("COCO_PATH")
    if env_path:
        return _runtime_from_candidate(Path(env_path), "COCO_PATH")

    on_path = shutil.which("coco")
    if on_path:
        return CocoRuntime(binary_path=on_path, source="PATH")

    for candidate in _common_install_paths():
        if candidate.is_file() and os.access(candidate, os.X_OK):
            return CocoRuntime(binary_path=str(candidate), source="common_path")

    raise CLINotFoundError(
        "coco binary not found. Install it or set COCO_PATH environment variable."
    )


def find_coco_binary(
    binary_path: str | os.PathLike[str] | None = None,
    *,
    env: Mapping[str, str] | None = None,
) -> str:
    """Return only the resolved binary path for simple callers."""

    return resolve_coco_runtime(binary_path, env=env).binary_path


def _runtime_from_candidate(candidate: Path, source: str) -> CocoRuntime:
    if candidate.is_file() and os.access(candidate, os.X_OK):
        return CocoRuntime(binary_path=str(candidate), source=source)
    raise CLINotFoundError(f"coco binary not found at {candidate}")


def _common_install_paths() -> tuple[Path, ...]:
    return (
        Path.home() / ".cargo" / "bin" / "coco",
        Path("/usr/local/bin/coco"),
    )

File: coco-sdk/typescript/src/runtime.ts
import { accessSync, constants } from "node:fs";
import { delimiter, join } from "node:path";

import { CLINotFoundError } from "./errors.js";

export type CocoRuntime = {
  binaryPath: string;
  source: "explicit" | "COCO_PATH" | "PATH" | "common_path";
};

export function resolveCocoRuntime(
  binaryPath?: string | null,
  env: NodeJS.ProcessEnv = process.env,
): CocoRuntime {
  if (binaryPath) return runtimeFromCandidate(binaryPath, "explicit");

  if (env.COCO_PATH) return runtimeFromCandidate(env.COCO_PATH, "COCO_PATH");

  const onPath = findOnPath("coco", env.PATH ?? "");
  if (onPath) return { binaryPath: onPath, source: "PATH" };

  for (const candidate of commonInstallPaths()) {
    if (isExecutable(candidate)) return { binaryPath: candidate, source: "common_path" };
  }

  throw new CLINotFoundError("coco binary not found. Install it or set COCO_PATH.");
}

export function findCocoBinary(binaryPath?: string | null, env: NodeJS.ProcessEnv = process.env): string {
  return resolveCocoRuntime(binaryPath, env).binaryPath;
}

function runtimeFromCandidate(binaryPath: string, source: CocoRuntime["source"]): CocoRuntime {
  if (!isExecutable(binaryPath)) {
    throw new CLINotFoundError(`coco binary not found at ${binaryPath}`);
  }
  return { binaryPath, source };
}

function findOnPath(name: string, pathValue: string): string | null {
  for (const dir of pathValue.split(delimiter)) {
    if (!dir) continue;
    const candidate = join(dir, name);
    if (isExecutable(candidate)) return candidate;
  }
  return null;
}

function commonInstallPaths(): string[] {
  const home = process.env.HOME;
  return [
    ...(home ? [join(home, ".cargo", "bin", "coco")] : []),
    "/usr/local/bin/coco",
  ];
}

function isExecutable(path: string): boolean {
  try {
    accessSync(path, constants.X_OK);
    return true;
  } catch {
    return false;
  }
}
